BIP Draft: Testnet 5 (#2196)
bitcoin/bips: bip-XXXX.md

           Fabian Jahr <fjahr@protonmail.com>
  Status: Draft
  Type: Specification
  Assigned: ?
  License: CC0-1.0
  Discussion: 2026-06-02: https://groups.google.com/g/bitcoindev/c/kGUMTxOvdJA
```

## Abstract

A new test network with the goal of replacing [Testnet 4][BIP94]. Testnet 5 removes the difficulty
exception defined in Testnet 4. Sustained exploitation of this exception has made the network difficult
to use for testing. Additionally, Testnet 5 enforces the consensus rules specified in BIP 54 from block 1.

## Motivation

Testnet 4 included mitigations for an issue known as the [block storm attack][block-storms] which could render the
whole network unusable. This led to a depletion of block subsidies, which made it hard to acquire
coins for testing. However, Testnet 4 still retained a modified version of the difficulty exception rule
with the aim of allowing CPU users a limited path to acquire coins for testing, to mine non-standard
transactions that other miners would not relay, and to keep the chain moving if a large source of hash
power were to leave the network. Shortly after Testnet 4's introduction, the exception has been

> systematically and sustainably exploited, which prevented the exception from achieving the intended

Reviewer: I think you mean that the exploit has been sustained continuously, but I’m not sure I would call the it "sustainable". ;)

```suggestion
systematically and persistently exploited, which prevented the exception from achieving the intended
```

or maybe:

```suggestion
systematically and continuously exploited, which prevented the exception from achieving the intended
```

Author: Took "persistently" here.

numbers of blocks due to multiple difficulty-exception blocks competing for the tip. This led
to discussion about changing Testnet 4 to mitigate this issue (see [Bitcointalk][bitcointalk-thread]
for analysis and discussion).

In Testnet 5 there is no exception to the PoW rules. This appears to be the logical conclusion,
since any such exception could be exploited by a motivated attacker. This
ensures the network’s behavior matches mainnet as closely as possible.

BIP 54 is already enforced on signet through [Bitcoin Inquisition][signet-bip54] as of this BIP's creation.
However, signet does not allow miners to test that their software reliably follows the rules of
BIP 54. Testnet 5 provides a testing environment for this.

## Specification

Testnet 5 follows the same consensus rules as mainnet with the following two changes.

### BIP 54 activation

The rules specified in [BIP 54 version 1.0.0][BIP54] are active on Testnet 5 from block 1.

#### Problem Statement

BIP 54 proposes new consensus rules in order to fix several potential attack vectors. Namely 
it prevents the timewarp attack, reduces the worst-case block validation time, prevents Merkle

> tree weaknesses, and avoids duplicate transactions without [bip-0030][BIP30] validation.

Reviewer: It doesn’t "prevent" the Merkle tree weaknesses, but rather it doesn’t allow the weakness to be exploited. Perhaps:

```suggestion
it prevents the timewarp attack, reduces the worst-case block validation time, mitigates Merkle
tree weaknesses, and avoids duplicate transactions without [bip-0030][BIP30] validation.
```

Author: Done

   worth it for a Testnet that has the alternative option to start fresh, especially considering that
   the attackers exploiting Testnet 4's rules may use them to prevent a smooth deployment of new rules,
   including those of BIP 54.
2. Any fix to Testnet 4 that goes beyond a band-aid would require a hard fork, which brings its own
   implementation cost: adding hard-fork support to clients, handling the p2p layer correctly, etc.
   For a test network this isn't worth taking on, especially since a band-aid would only make
   Testnet 4 diverge further from mainnet.

Raising the minimum difficulty was dismissed for Testnet 4 because it would have prevented CPU
miners from utilizing the difficulty exception. With the exception removed in Testnet 5, this
argument no longer applies. A minimum difficulty of approximately 1,000,000 strikes a balance
between dampening the quick mining of large numbers of blocks shortly after launch, before the
difficulty has adjusted to the available hash power, and keeping a low barrier so that a handful
of at-home miners or a single ASIC can keep the network usable.

> ## Network Parameters

Reviewer: Aren’t the network parameters part of the specification?

Author: I moved the Network Parameters section into Specification now

### Consensus Rules

All consensus rules active on mainnet as of May 2026 are enforced from block 1, the most
recent of these being the Taproot softfork. Additionally, the rules of BIP 54 are enforced
from block 1, as specified above.

### Genesis Block

TODO: Mine the block. The values below are placeholders inherited from Testnet 4. Notes
for the miner:

> - For the `Pubkey` field, use a recent Bitcoin mainnet block hash. This single field then

Reviewer: "`Pubkey` field" has not been introduced here?

Author: Added something of a clarification on this (see other comment).

> - For the `Pubkey` field, use a recent Bitcoin mainnet block hash. This single field then

Reviewer: Perhaps add a line along the lines of the following to give the reader a sense what you’re about to talk about:

```suggestion
The following information is included in the Genesis block coinbase transaction’s coinbase field:

- For the `Pubkey` field, use a recent Bitcoin mainnet block hash. This single field then
```

Author: I introduced something similar to clarify, but this part should be mostly removed when the actual Genesis block has been mined.

Reviewer: It could perhaps be turned into a footnote to explain the methodology in case someone else wants to start the next testnet in a couple years.

Author: Ok, using a footnote now and adding the intention for the methodology which is a bit slimmer so it's good to keep around after the block has been added.

- `Time stamp` time of mining or that of the block which hash is used in `Pubkey`
- `nBits`: `0x1a0fffff`, per the raised minimum difficulty.
- `Version: 1`

Testnet 4 placeholders:

> * Message: <code>03/May/2024 000000000000000000001ebd58c244970b3aa9d783bb001011fbe8ea8e98e00e</code>
> * Pubkey: <code>000000000000000000000000000000000000000000000000000000000000000000</code>
> * Time stamp: 1714777860
> * Nonce: 393743547
> * Difficulty: <code>0x1d00ffff</code>
> * Version: 1
> 
> The resulting Genesis block hash is <code>00000000da84f2bafbbc53dee25a72ae507ff4914b867c565be350b0da8bf043</code>, and the block hex is <code>0100000000000000000000000000000000000000000000000000000000000000000000004e7b2b9128fe0291db0693af2ae418b767e657cd407e80cb1434221eaea7a07a046f3566ffff001dbb0c78170101000000010000000000000000000000000000000000000000000000000000000000000000ffffffff5504ffff001d01044c4c30332f4d61792f323032342030303030303030303030303030303030303030303165626435386332343439373062336161396437383362623030313031316662653865613865393865303065ffffffff0100f2052a010000002321000000000000000000000000000000000000000000000000000000000000000000ac00000000</code>.

### Message Start

> The message start is defined as <code>0x46495645</code>. These four bytes spell `FIVE` when

Reviewer: Do you mean the “network magic”?

Author: Yes, in Testnet 4 it is called Message Start which is the term used in the chainparams in Bitcoin Core. But I made a little change to also mention "network magic".

interpreted as ASCII.

### Network Parameters

The default p2p port for Testnet 5 is `18335`.

## Backward Compatibility

Testnet 5's consensus rules are not compatible with those of Testnet 3 and Testnet 4. The
consensus rules differ in both directions: Testnet 5 enforces the BIP 54 consensus rules from
block 1 which is not the case for Testnet 3 or Testnet 4. Testnet 5 also does not apply the

> difficulty exception rule from Testnet 3 or Testnet 4 requires.

Reviewer: Maybe also make it clear for newer readers:
They use different genesis blocks, so their UTXO sets will be unique, and it is infeasible for any transactions to ever be replayable across these networks. Since output scripts look are compatible, addresses could be reused across these networks.

Author: Sure, I added a paragraph clarifying this.

```
  BIP: ?
  Layer: Applications
  Title: Testnet 5
  Authors: Pol Espinasa <polespinasa@protonmail.com>
           Fabian Jahr <fjahr@protonmail.com>
  Status: Draft
  Type: Specification
  Assigned: ?
  License: CC0-1.0
  Discussion: 2026-06-02: https://groups.google.com/g/bitcoindev/c/kGUMTxOvdJA
  Version: 0.1.0
```

## Abstract

> A new test network with the goal of replacing [Testnet 4][BIP94]. Testnet 5 removes the difficulty

Reviewer: Nit: The abstract starts on a sentence fragment without a verb.

Author: Changed

exception defined in Testnet 4. Sustained exploitation of this exception has made the network difficult
to use for testing. Additionally, Testnet 5 enforces the consensus rules specified in BIP 54 from block 1.

## Motivation

Testnet 4 included mitigations for an issue known as the [block storm attack][block-storms] which could render the

> whole network unusable. This led to a depletion of block subsidies, which made it hard to acquire

Reviewer: It’s ambiguous whether "this" refers to the mitigations or the block storms.

```suggestion
whole network unusable. Block storm attacks led to a depletion of block subsidies, which made it hard to acquire
```

Author: Fixed

> coins for testing. However, Testnet 4 still retained a modified version of the difficulty exception rule

Reviewer: Since much of this paragraph discusses the issues with the difficulty exception, perhaps it should be briefly explained.

Author: I was thinking that people could hop over to BIP94 for the details but I added another sentence to give a short explainer.

```

>   BIP: ?

Reviewer: ```suggestion
  BIP: 95
```

Author: Done

  Layer: Applications
  Title: Testnet 5
  Authors: Pol Espinasa <polespinasa@protonmail.com>
           Fabian Jahr <fjahr@protonmail.com>
  Status: Draft
  Type: Specification

>   Assigned: ?

Reviewer: ```suggestion
  Assigned: 2026-06-22
```

Author: Done
